### ISSUE TITLE
Client components missing in SSR manifest when imported from files not in `webpackRscLayerName`

### ISSUE THREAD

**Issue by quentinmit** [unstubbable/mfng#71]
It's unclear to me how I am supposed to configure the `resource` regexes for the `webpackRscLayerName`. It seems like any imports of client components need to be seen by the `rscServerLoader`, which is only used by those layers. If I reference a client component from a file not included in that regex, I end up with two copies of the client component in my bundle, one of them having client references and one having the original code (despite being in the react-server layer).  This also causes the resulting `clientReferencesMap` to not have `ssrId` populated, because webpack will append a random string to the module ID.

Is the idea that `webpackRscLayerName` should have all files that might include client components? It seems like that fails if other client components are also included in the layer.

Basically any guidance on which source modules should be passed to each of the loaders/plugins would be very appreciated.

**Comment by quentinmit** [comment]
Sure, I tried to make one at https://github.com/quentinmit/mfng-repro

You can build and launch it with `npm run start`, and then `curl http://localhost:3000/` to trigger a render.

Take a look at the three different options I've described in `node-webpack.config.mjs`; all three of those regex sets fail with different errors.

Can you help me figure out what the right combination of webpack layers is?

**Comment by unstubbable** [comment]
Thanks for the repro! The main server component needs to live in its own file so that we can target it in the webpack RSC layer rule. See https://github.com/quentinmit/mfng-repro/pull/1

**Comment by quentinmit** [comment]
How does that work if there's more than one "main server component"?

**Comment by unstubbable** [comment]
As you can see in your webpack config, `resource` accepts an array.

**Comment by quentinmit** [comment]
I think there's still something missing in the webpack config. After merging your PR into my mfng-repro repo, and trying to add a `Link` component into `App.tsx`, , compiles now fail with

```
  ERROR in ./node_modules/@mfng/core/lib/client/use-router.js 3:29-48
  export 'createContext' (imported as 'React') was not found in 'react' (possible exports: Children, Fragment, Profiler, StrictMode, Suspense, __SERVER_INTERNALS_DO_NOT_USE_OR_WARN_USERS_THEY_CANNOT_UPGRADE, cache, cloneElement, createElement, createRef, forwardRef, isValidElement, lazy, memo, startTransition, use, useActionState, useCallback, useDebugValue, useId, useMemo, version)
   @ ./node_modules/@mfng/core/lib/client/index.js 3:0-32 3:0-32
   @ ./src/components/App.tsx 4:0-41 6:150-154
   @ ./src/routes.tsx 12:0-39 38:68-71
   @ ./src/app.ts 4:0-39 11:21-32
   @ ./src/main.ts 1:0-24 3:0-3
```

which I think means it's trying to use the server version of React despite the `"use client"` at the top of `link.tsx`.

I pushed this latest version to the `mfng-repro` repo if you wouldn't mind taking another look.

P.S. Is there an IRC/Matrix/etc. channel for mfng? I feel bad cluttering up the Issues with "am I holding it right?" questions.

**Comment by unstubbable** [comment]
Thanks again for raising these issues! I've implemented fixes and integrated them in https://github.com/quentinmit/mfng-repro/pull/2.

**Comment by quentinmit** [comment]
Thanks so much! I'm afraid I've found two new issues:

- `"use client"` is ignored when using a default export (e.g. `import Tooltip from "react-bootstrap/Tooltip";` fails unless it's imported from another file that also has `"use client"`)

I pushed a demo of this to the mfng-repro repo. I think it also fails when using a class component, but I didn't make a repro of that yet.

- webpack race conditions I haven't managed to track down

Depending on the `parallelism` setting, I'm seeing failures of the form:

```
  ERROR in ./node_modules/prop-types/lib/has.js
  Cannot read properties of undefined (reading 'length')
  TypeError: Cannot read properties of undefined (reading 'length')
      at WasmHash._updateWithBuffer (/home/user/repo/node_modules/webpack/lib/util/hash/wasm-hash.js:108:23)
      at WasmHash.update (/home/user/repo/node_modules/webpack/lib/util/hash/wasm-hash.js:51:8)
      at BatchedHash.update (/home/user/repo/node_modules/webpack/lib/util/hash/BatchedHash.js:53:14)
      at NormalModule.updateHash (/home/user/repo/node_modules/webpack/lib/NormalModule.js:1521:8)
      at Compilation._createModuleHash (/home/user/repo/node_modules/webpack/lib/Compilation.js:4107:11)
      at Compilation.createModuleHashes (/home/user/repo/node_modules/webpack/lib/Compilation.js:4052:25)
      at /home/user/repo/node_modules/webpack/lib/Compilation.js:3127:11
      at Hook.eval [as callAsync] (eval at create (/home/user/repo/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
      at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (/home/user/repo/node_modules/tapable/lib/Hook.js:18:14)
      at /home/user/repo/node_modules/webpack/lib/Compilation.js:3083:36
   @ ./node_modules/prop-types/factoryWithTypeCheckers.js 14:10-30
   @ ./node_modules/prop-types/index.js 14:19-55
   @ ./node_modules/react-bootstrap/esm/OverlayTrigger.js 4:0-35 37:20-35
   @ ./src/App.tsx 6:0-60 134:32-46
```

(Note that what's undefined is `this.buildInfo.hash` in the `NormalModule` object that represents `./node_modules/prop-types/lib/has.js`)

I haven't been able to simplify this down to something that I can reproduce in the mfng-repro repo though, so it's possible that this has nothing to do with the mfng plugins.

**Comment by quentinmit** [comment]
I was able to make a small reproducer for the webpack race condition. It's at https://github.com/quentinmit/mfng-repro/tree/webpack-race, and it fails to compile most of the times I try (but occasionally it succeeds). Based on the different output from each run, it seems like there's a race condition with a data structure inside webpack.

**Comment by unstubbable** [comment]
Thanks again for the repro, that helps a lot! Both of these issues should be fixed with `@mfng/webpack-rsc@4.2.1`. Let me know if you find anything else.

**Comment by quentinmit** [comment]
> Thanks again for the repro, that helps a lot! Both of these issues should be fixed with `@mfng/webpack-rsc@4.2.1`. Let me know if you find anything else.

I filed https://github.com/unstubbable/mfng/issues/80 for another manifest generation bug we encountered.
